src/fromager/sources.py

```diff
@@ -123,8 +124,10 @@ def resolve_dist(
         logger.debug(f"{req.name}: could not resolve {req} with {constraint}: {err}")
         raise
 
-    candidate: resolver.Candidate
+    candidate: Candidate
     for candidate in result.mapping.values():
+        if candidate not in ctx.resolver_cache[candidate.identifier]:
```

Comment:
Having to remember to check this from the outside of the cache before adding something makes me think we do want a cache class so we can have an `add()` method that hides the check.